### ISSUE TITLE
v1.0.0 needs to be issued

### ISSUE THREAD

**Issue by jsumners** [fastify/fastify#229]
With the GitHub project fast approaching 2,500 stars, people need to be able to have confidence that the library they are using is stable. Looking at the roadmap today (https://github.com/fastify/fastify/projects/1) I can see that a couple new issues were added (#135 and #121), but I think the authentication issue has been solved (https://github.com/fastify/fastify-auth and https://github.com/fastify/fastify-bearer-auth) well enough for 1.0.0.

In my opinion, @mcollina should be able to speak about 1.0.0 in his presentation at the beginning of October.

cc: @fastify/benchmarks 

2017-09-24: I have segregated this list into "blocker" and "non-blockers". A non-blocker is an issue that we would like to have solved before 1.0.0, but isn't completely necessary. ~ jsumners
___
Todo (blockers):
- [x] 404/500 routes (https://github.com/fastify/fastify/issues/135)
- [x] http caching headers (#121)
- [x] Add headers to validation schema (#191)
- [x] Add headers to `request` object (https://github.com/fastify/fastify/issues/200)
- [x] Take a final decision in https://github.com/fastify/fastify/issues/169
- [x] `onSend` hook (#291)
- [x] Fix printing stack traces https://github.com/fastify/fastify/issues/258
- [x] Allow payload to be modified in `onSend` (#447)
- [x] Fix `this` context on hooks (#120)
- [x] Remove `reply.send(promise)` (https://github.com/fastify/fastify/issues/448)
- [x] setNotFoundHandler is suppressing duplicated routes checking #451
- [x] Remove callback from `register` #460
- [x] Support metadata in `fastify-plugin` #456  #466 
- [x] Update content parser API #525 
- [x] avvio says it's already booted before `fastify.listen` is invoked #491 #561
- [x] Drop Boom support #542 #558
- [x] Use `http-errors` for generating errors #543 #571

Todo (non-blockers):
- [ ] Refactor documentation to separate concerns (intro vs reference) (#296)
- [x] Finish off fastify-swagger https://github.com/fastify/fastify-swagger/pull/1

**Comment by delvedor** [comment]
Features that we must implement before 1.0.0:
- 404/500 routes (https://github.com/fastify/fastify/issues/135)
- http caching headers (#121)
- Add headers to validation schema (#191)
- Add headers to `request` object (https://github.com/fastify/fastify/issues/200)
- Finish the implementation of [`fastify-swagger`](https://github.com/fastify/fastify-swagger)
- Take a final decision in https://github.com/fastify/fastify/issues/169

**Comment by jsumners** [comment]
@delvedor you should have permissions to be able to edit the post. I am completely fine with that.

**Comment by delvedor** [comment]
> @delvedor you should have permissions to be able to edit the post. I am completely fine with that.

Done :)

**Comment by cagataycali** [comment]
For first stable version, this proposals looks good for me! 🎉

**Comment by jsumners** [comment]
I disagree with #169 being a blocker for 1.0.0. That is clearly a divisive proposal and unlikely to be settled any time soon.

**Comment by delvedor** [comment]
> I disagree with #169 being a blocker for 1.0.0. That is clearly a divisive proposal and unlikely to be settled any time soon.

Yes, and since it is a *divisive proposal* I think we should take a final decision on that.
If we keep `req, reply` we should provide a better way to help user interact with both `request` and `reply` at the same time when creating a plugin.

**Comment by StarpTech** [comment]
I would add this to the list https://github.com/fastify/fastify/issues/258

**Comment by mcollina** [comment]
@StarpTech done.

**Comment by StarpTech** [comment]
Hi, why is swagger a blocker for a major release?

**Comment by jsumners** [comment]
@StarpTech for me, it's not. But I understand why it is. Fastify's goal is for writing RESTful API servers. Thus, it would be good to be able to support documenting those APIs. I think things like Swagger are way more trouble than their worth (the code for Swagger to read ends up being more than the actual method), but there's a lot of people that like it.

**Comment by StarpTech** [comment]
Ok, I understand it but it's still a plugin and not part of the core. Is anybody working on it? Should we discuss some proposals about the integration with fastify e.g configuration approach vs comments?

**Comment by mcollina** [comment]
The actual problem for 1.0.0 is that we had semver-major changes in the last 5 minor releases. I think our API has not settled yet.

**Comment by mcollina** [comment]
I propose we drop fastify-swagger from the list. We can ship it later. 404 and 500 have already a PR open.. is there anyone that wants to add support for the http caching headers in Request and Reply, similar to what Hapi does?

**Comment by jsumners** [comment]
+1 to Swagger not being a blocker.

I'm not entirely sure what needs to be done to tackle #121.

**Comment by allevo** [comment]
I'd like to add https://github.com/fastify/fastify/issues/296 to this list

**Comment by jsumners** [comment]
@allevo done.

**Comment by StarpTech** [comment]
Http-Caching is done? https://github.com/fastify/fastify/issues/121

**Comment by jsumners** [comment]
No. It just has its own repository.

**Comment by StarpTech** [comment]
But what's the milestone in caching?

**Comment by jsumners** [comment]
There is discussion about it in the issue tracker on that repo. But the short of it is that I am currently running for office in my city and don't have free time to work on it. Thankfully, the final Election Day is November 7.

**Comment by jsumners** [comment]
Blockers list updated with closing of `onSend` and addition of two new items.

**Comment by mcollina** [comment]
Thanks!

**Comment by jsumners** [comment]
@fastify/fastify please re-review the blockers list. There is only one open at this time. Make sure there are not any others that should be there.

**Comment by allevo** [comment]
I'd like to finish the discussion https://github.com/fastify/fastify/issues/169

**Comment by jsumners** [comment]
Accidental close. Reopened.

**Comment by delvedor** [comment]
[`fastify-swagger`](https://github.com/fastify/fastify-swagger) has been finished.
It probably needs some work, but the initial release is ready :)

**Comment by jsumners** [comment]
@fastify/fastify please give this comment a 👍 or 👎 to vote for 1) *accepting* #169 and 2) requiring #169 as a blocker for v1.0.0.

👍 => Accept #169 and implement it before issuing v1.0.0
👎 => Reject #169, at least for v1.0.0

Please complete this vote by 10 November 2017 @ 17:00 UTC.

**Comment by mcollina** [comment]
I've added https://github.com/fastify/fastify/issues/448 to the blockers.

**Comment by jsumners** [comment]
With 1 vote for and 3 votes against, we will not be blocking on #169. The context issue can be re-addressed for a v2.0.0.

![fastify](https://user-images.githubusercontent.com/321201/32670911-4c554c22-c613-11e7-929b-44a1e3b1df8a.png)

**Comment by delvedor** [comment]
Added #451 to the blockers.

**Comment by jsumners** [comment]
Added #456 and #460 to the blockers.

**Comment by jsumners** [comment]
We need to have a cutoff date for new issues to be added to this list. When the date is reached, the API shall be considered shippable for 1.0.0 and no new issues may be added that can alter it. @fastify/fastify (and community at large) please vote on this comment by 19 November 2017, 12:00 UTC.

🎉 => cutoff date of 01 December 2017, 00:00 UTC
👍 => cutoff date of 16 December 2017, 00:00 UTC

These dates should allow a v1.0.0 to be released by year end.

**Comment by delvedor** [comment]
>  16 December 2017, 00:00 UTC

Basically one month from today, that's fine to me.

**Comment by mcollina** [comment]
👍 with the cutoff, I'll be traveling for most of the period so I'm actually abstaining. To me it's like it's next week :).

**Comment by jsumners** [comment]
Note: I also abstain from the polls I create.

**Comment by jsumners** [comment]
With a tie, I couldn't think of a fairer way to decide it, so I rolled a die -- https://youtu.be/wwGRnAwnsNI

Result: cutoff date is 16 December 2017, 00:00 UTC.

**Comment by coreyfarrell** [comment]
Could these issues be tagged with a label or milestone?  That is how I would normally find such tickets.

In addition would new labels `semver-minor` / `semver-major` be considered?  Once the `semver-major` label is applied it would be easier to know that a patch is forbidden from v1.x and control how a patch is merged.  It would also be useful in the release notes, see node.js (breaking changes are clearly labeled SEMVER-MAJOR).

**Comment by delvedor** [comment]
@coreyfarrell hi! Thank you for the suggestions!
I agree for the `semver-*` label, I'll create them, and once the `v1.0.0` is done we will start using them :)

**Comment by mcollina** [comment]
Ping, cutoff data is 16th of December.

**Comment by jsumners** [comment]
@fastify/fastify reminder: you have 3 days to submit new API changes that will block the release of v1.0.0.

**Comment by mcollina** [comment]
Cutoff is happening tomorrow. Does anybody has anything more to add?

**Comment by jsumners** [comment]
To be clear: the is 00:00 UTC. That is 8 hours from the time of this comment.

**Comment by allevo** [comment]
https://github.com/mcollina/avvio/pull/36 and https://github.com/mcollina/avvio/pull/37 could impact the release.

**Comment by jsumners** [comment]
Please create an issue to track the end result.

**Comment by mcollina** [comment]
@allevo those two issues are already tracked in this list.

**Comment by jsumners** [comment]
Just one more item on the list 😄

**Comment by delvedor** [comment]
Should https://github.com/fastify/fastify/issues/296 be a blocker?

**Comment by jsumners** [comment]
@delvedor No. Documentation is easily revised as semver minor/patch. Updating the documentation doesn't actually change the API.

**Comment by delvedor** [comment]
@jsumners you are right, but I'm proposing to make it so because it is our first major release, I think that have everything well done and documented is important.

**Comment by Ethan-Arrowood** [comment]
If there is one thing I've learned from being in the Node.JS & React.JS space is documentation is a big ticket item. I'd be happy to help write, proofread, and review 👍

**Comment by mcollina** [comment]
@Ethan-Arrowood I agree with that statement! If you would like to do a big PR it would be fantastic.

0.37 is feature complete for 1.0, we just need to a) remove stuff and b) write better docs (optionally).

If anyone wants to give a stab at #460, it needs to be done in avvio. It's going to be tricky. I'm going to be away from a laptop until Jan 2nd.

**Comment by jsumners** [comment]
@delvedor I understand. But I believe earlier in this thread we came to the consensus that the documentation is serviceable for v1.0.0. 

That being said, I have the next week off work. If @Ethan-Arrowood doesn't tackle the problem, I will dig into it some time next week.

**Comment by mcollina** [comment]
I think it’s workable. There are also a couple more things I would like to do for 1.0 on our website, and maybe have a launch blog post.

Also, we’d have to update our benchmarks as well.

A lot of these are blocked by me, so feel free to take any.

**Comment by Ethan-Arrowood** [comment]
@mcollina @jsumners I'll organize a nice issue post for the documentation later this evening and then start working on it as I can!

**Comment by mcollina** [comment]
After https://github.com/fastify/fastify/pull/651, we are ready to go.

I think we should come up with rc candidates now for a couple of weeks, and then release?

**Comment by delvedor** [comment]
> I think we should come up with rc candidates now for a couple of weeks, and then release?

👍

**Comment by jsumners** [comment]
Agreed on a couple RC releases. I'm also going to make an attempt at resolving my issue with options. Hopefully within the next day or two.

**Comment by allevo** [comment]
I'd like to wait till a nicer solution to avoid https://github.com/mcollina/avvio/pull/36#issuecomment-357260628 is provided.

**Comment by jsumners** [comment]
Are we ready for rc1?

**Comment by Mat-thieu** [comment]
So if I understand this thread correctly, v1.0.0 is basically ready but until the docs are updated it won't be released?

**Comment by Ethan-Arrowood** [comment]
@Mat-thieu I'm pretty sure docs are nonblocking but I could be wrong

**Comment by Mat-thieu** [comment]
@Ethan-Arrowood According to the list they aren't, picked up this framework today and I really like it, but I'm afraid to use it in production at this version so I was wondering what the status is

**Comment by jsumners** [comment]
The docs refactor would be a cherry on top of a v1.0.0 release. But, unless any other @fastify/fastify  member says otherwise, we are ready to cut a v1.0.0-rc1 release. Which is to say, the current master (and published version) is what you should expect in v1.0.0.

**Comment by Mat-thieu** [comment]
@jsumners Sweet!

**Comment by mcollina** [comment]
There are a few breaking changes in master, plus a couple of outstanding PRs.
I would issue another minor with what we have on master + outstanding PRs, and then some RCs.
We should figure out some LTS strategy.

**Comment by Ethan-Arrowood** [comment]
Something I discovered recently is that travis.ci lets you have build stages now so we could run the unit tests in multiple version of node in order to maintain LTS

**Comment by jsumners** [comment]
It's clear in #709 that we definitely need to get this issue resolved. It looks like we have one bug outstanding (#693), and it doesn't seem to be affecting many people. I think we should be issuing an rc release promptly.

**Comment by mcollina** [comment]
@jsumners that should be solved in the RC phase.

**Comment by jsumners** [comment]
I think v1.0.0-rc1 should be released this week. What say you @fastify/fastify ?

**Comment by allevo** [comment]
@jsumners A breaking-change issue is opened #718. #727 is another breaking-change.

**Comment by delvedor** [comment]
I'd love to release a rc as soon as possible, but currently we have two breaking changes proposed. That's too much. I think that if in a week we do not found anything that could cause a breaking change, then we can release an rc and consider the api stable enough.

**Comment by allevo** [comment]
> I think that if in a week we do not found anything that could cause a breaking change

Apparently, it seems a bit difficult 😃

**Comment by delvedor** [comment]
@allevo that's not a problem, we do not have any fixed date, we just want to deliver a good and stable api before release v1 and I think no one can complain with this.

**Comment by jsumners** [comment]
The issue is that we can come up with a breaking change every week. We, as programmers, will always see a better way to do a thing we have already done and seek to change it. No, there isn't a date set for release (I'd argue that there *should* be one), but as has been shown in the past couple of weeks, there are people already relying on Fastify with its current API. We need to lock it down *very* soon or risk losing the traction already gained.

I think the current PRs labeled as breaking can be resolved in short order. That's why I think an rc1 can be issued this week.

**Comment by Ethan-Arrowood** [comment]
I don't know sort of plans you have for releasing 1.0.0, but if we are interested in publishing a blog post I have a good relationship with the FreeCodeCamp publication on Medium and could help facilitate a release article. (Would be happy to help write it as well).

**Comment by mcollina** [comment]
I think we can do as many rcs as we like. Let’s target next monday.

**Comment by delvedor** [comment]
The first release candidate is out! 🎉 
https://github.com/fastify/fastify/releases/tag/v1.0.0-rc.1

**Comment by mcollina** [comment]
Fastify has come a long way! Should we pick a date and release then? I do not want to keep the rc going for more than 2-3 weeks.

**Comment by jsumners** [comment]
March 1?

**Comment by mcollina** [comment]
I like March 1!

**Comment by jsumners** [comment]
Looks like people have voted that March 1 be the v1.0.0 target date.

**Comment by yamalight** [comment]
@jsumners maybe it's also worth releasing RC versions for all official extensions? 
at least some of them are currently [incompatible](https://github.com/fastify/fastify-auth/issues/9) with 1.0-rc1

**Comment by jsumners** [comment]
@yamalight it's up to the maintainers of the plugins to do that. I released updates yesterday for the ones I maintain. PRs to currently incompatible plugins would be helpful. Basically, they need:

```js
module.exports = fp(plugin, {
  fastify: '>=1.0.0-rc.1',
  name: 'foo-plugin'
})
```

And after the rc phase, they should change the minimum version to `^1.0.0`.

**Comment by Ethan-Arrowood** [comment]
@jsumners I'll make some prs against the JWT libraries later today

**Comment by allevo** [comment]
I'd like to wait for https://github.com/delvedor/find-my-way/issues/60

**Comment by delvedor** [comment]
If someone wants to give a try to that issue be my guest!
I'll give it an eye asap, but at the moment I'm stuck at work.

**Comment by delvedor** [comment]
The router bug has been fixed.
The `v1.0.0` release is planned for the 5th of march! *(unless very serious bugs comes out)*

**Comment by delvedor** [comment]
I guess we can close this one :)

https://medium.com/@fastifyjs/fastify-goes-lts-with-1-0-0-911112c64752
